refactor: make CTxMemPool::nCheckFrequency a const and a constructor argument

ellemouton opened the pull request:
This PR refactors CTxMemPool::nCheckFrequency to be a const. This can be
done since the CTxMemPool object is no longer a global variable and so
the value of nCheckFrequency is now known at the time that the object is
constructed and so can be passed into the CTxMemPool constructor. This
also means that the nCheckFrequency variable no longer needs to be
guarded by the 'cs' lock.

ellemouton commented on src/txmempool.cpp:
```diff
@@ -331,15 +331,10 @@ void CTxMemPoolEntry::UpdateAncestorState(int64_t modifySize, CAmount modifyFee,
 {
     _clear(); //lock free clear
-
-    // Sanity checks off by default for performance, because otherwise
-    // accepting transactions becomes O(N^2) where N is the number
-    // of transactions in the pool
```
i feel like this comment should still be somewhere...?

ellemouton commented on src/init.cpp:
```diff
@@ -1389,16 +1389,16 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
+    int ratio = std::min<int>(std::max<int>(args.GetArg("-checkmempool", chainparams.DefaultConsistencyChecks() ? 1 : 0), 0), 1000000);
+    uint32_t checkFrequency = 0;
+    if (ratio != 0) {
+    	checkFrequency = static_cast<uint32_t>((1.0/ratio) * 4294967295.0);		
```
im not sure where this magic number comes from...? (copied from src/txmempool.h 'setSanityCheck'). I feel like there should be a comment explaining where it comes from

jnewbery replied:
4294967295 is one less than the maximum uint32 (2^32 = 4294967296). The value was originally the max uint32_t, but was changed to be one less to fix an overflow bug: https://github.com/bitcoin/bitcoin/commit/e9e616323bc44daf61f4358f5b2de58e90b5c602

jnewbery replied:
You might be able to replace it with (UINT32_MAX - 1) (see https://en.cppreference.com/w/cpp/header/cstdint)

jnewbery replied:
You could also pass the ration int into the constructor and do this int -> float -> uint32_t conversion inside the constructor.

jnewbery commented on src/test/util/setup_common.cpp:
```diff
@@ -142,7 +142,6 @@ TestingSetup::TestingSetup(const std::string& chainName, const std::vector<const
     pblocktree.reset(new CBlockTreeDB(1 << 20, true));
 
     m_node.mempool = MakeUnique<CTxMemPool>(&::feeEstimator);
-    m_node.mempool->setSanityCheck(1.0);
```
You need to update the constructor call to include the checkFrequency 1 here.

jnewbery commented on src/txmempool.h:
```diff
@@ -602,7 +602,7 @@ class CTxMemPool
     /** Create a new CTxMemPool.
      */
-    explicit CTxMemPool(CBlockPolicyEstimator* estimator = nullptr);
+    explicit CTxMemPool(CBlockPolicyEstimator* estimator = nullptr, uint32_t checkFrequency = 0);
```
Local variables and arguments names should use camel_case. See our developer notes for all our project style guidelines: https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md

jnewbery commented on src/init.cpp:
```diff
@@ -1389,16 +1389,19 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
+    // Sanity checks off by default for performance, because otherwise
+    // accepting transactions becomes O(N^2) where N is the number
+    // of transactions in the pool
+    uint32_t checkFrequency = 0;
```
Use camel_case here.

jnewbery replied:
You can use braced initialization here:

```suggestion
    uint32_t checkFrequency{0};
```

jnewbery commented on src/init.cpp:
```diff
@@ -1389,16 +1389,19 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
     assert(!node.connman);
     node.connman = MakeUnique<CConnman>(GetRand(std::numeric_limits<uint64_t>::max()), GetRand(std::numeric_limits<uint64_t>::max()), args.GetBoolArg("-networkactive", true));
 
+    // Sanity checks off by default for performance, because otherwise
```
Good idea to keep this comment. Perhaps move it to the comment for `nCheckFrequency` in txmempool.h

jnewbery reviewed (commented):
Looks great. Most of my comments are style nits, but I think you've changed the behaviour in the tests in one place.

ellemouton committed:
refactor: Avoid double to int cast for nCheckFrequency

Use a ratio instead of a frequency that requires a double to int cast
for determining how often a mempool sanity check should run.

ellemouton committed:
refactor: Make CTxMemPool::m_check_ratio a const and a constructor argument

Since m_check_ratio is only set once and since the CTxMemPool object is
no longer a global variable, m_check_ratio can be passed into the
constructor of CTxMemPool. Since it is only read from after
initialization, m_check_ratio can also be made a const and hence no
longer needs to be guarded by the cs mutex.

ellemouton committed:
refactor: Clean up CTxMemPool initializer list

Shorten the CTxMemPool initializer list using default initialization
for members that dont depend on the constuctor parameters.

jnewbery commented on src/txmempool.h:
```diff
@@ -488,8 +488,8 @@ class SaltedTxidHasher
 private:
-    uint32_t nCheckFrequency GUARDED_BY(cs); //!< Value n means that n times in 2^32 we check.
-    std::atomic<unsigned int> nTransactionsUpdated; //!< Used by getblocktemplate to trigger CreateNewBlock() invocation
+    const int m_check_ratio; //!< Value n means that 1 times in n we check.
```
I suggest you make this change in the first commit (_refactor: Avoid double to int cast for nCheckFrequency_), since that's where the meaning of this changes from a frequency to a ratio.

jnewbery commented on src/txmempool.h:
```diff
@@ -599,10 +599,13 @@ class CTxMemPool
+    /** Create a new CTxMemPool. 
+     * Sanity checks will be off by default for performance, because otherwise 
+     * accepting transactions becomes O(N^2) where N is the number of transactions 
+     * in the pool.
```
Good idea to move this comment here. You could use doxygen-style comments for these params (e.g. here's an example: https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.h#L63-L66). That helps doxygen parsers generate documentation (e.g. here: https://doxygen.bitcoincore.org/class_peer_manager.html#a28fb5cd394936a42d9e68eba2e521677)

jnewbery commented on src/txmempool.cpp:
```diff
@@ -619,13 +614,13 @@ static void CheckInputsAndUpdateCoins(const CTransaction& tx, CCoinsViewCache& m
 {
-    LOCK(cs);
-    if (nCheckFrequency == 0)
+    if (m_check_ratio == 0)
```
When touching lines of code, feel free to fix them up to use the modern code style guidelines. In new code, we don't allow `if` statements to have single indented lines without braces, since that can lead to bugs like https://www.imperialviolet.org/2014/02/22/applebug.html - people see some indented if code, add an extra line to it and don't realise that the new code is always executed. This line would become:

```
if (m_check_ratio == 0) return;
```

or:

```
if (m_check_ratio == 0) {
    return;
}
```

Both are fine. If the code is short, I prefer the first for concision.

jnewbery commented on src/txmempool.cpp:
```diff
@@ -619,13 +614,13 @@ static void CheckInputsAndUpdateCoins(const CTransaction& tx, CCoinsViewCache& m
+    if (GetRand(m_check_ratio) >= 1)
         return;
 
+    LOCK(cs);
```
You can move this even further down, to just above the `CCoinsViewCache mempoolDuplicate` declaration, so we're not logging while holding the lock.

All of these things are very marginal, but minimizing where locks are held is generally good practice.

ellemouton replied:
I put the lock here cause the Log line uses the 'mapNextTx' variable which i think is guarded by the 'cs' lock. (src/txmempool.h:600). Which is why i grabbed the lock before the log line. is that right?

jnewbery replied:
Yes, you're right. Ignore this!

jnewbery commented on src/init.cpp:
```diff
@@ -1389,16 +1389,12 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
     assert(!node.mempool);
-    node.mempool = MakeUnique<CTxMemPool>(&::feeEstimator);
-    if (node.mempool) {
-        int ratio = std::min<int>(std::max<int>(args.GetArg("-checkmempool", chainparams.DefaultConsistencyChecks() ? 1 : 0), 0), 1000000);
-        if (ratio != 0) {
-            node.mempool->setSanityCheck(1.0 / ratio);
-        }
-    }
+    node.mempool = MakeUnique<CTxMemPool>(&::feeEstimator, check_ratio);
```
This is a bit messy now. I think it would be tidier as:

```
    assert(!node.mempool);
    int check_ratio = std::min<int>(std::max<int>(args.GetArg("-checkmempool", chainparams.DefaultConsistencyChecks() ? 1 : 0), 0), 1000000);
    node.mempool = MakeUnique<CTxMemPool>(&::feeEstimator, check_ratio);
```

(The comment is outdated - it's from before the mempool was constructed here, and refers to adding a raw mempool pointer to the node context).

jnewbery reviewed (commented):
This is basically ready. I've left some very nitty comments. Feel free to take or leave them. Once you've done that, I think this is ready to PR to the main repo :)

jnewbery commented:
Oh, one more thing - you could make the git commit logs just a little more verbose. No need to write an essay, but just a short justification for the changes.

jnewbery commented:
New commit logs look good. One small typo:

s/Use a ratio instead on/Use a ratio instead of/

This is definitely ready to PR!

jnewbery commented:
You can close this now :)
